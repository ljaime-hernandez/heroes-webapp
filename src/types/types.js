
// the types contain a string in this scenario just so they have information
// but the content is irrelevant for this example
export const types = {
    login: '[auth] login',
    logout: '[auth] logout'
}
