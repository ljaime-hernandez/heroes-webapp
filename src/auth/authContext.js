import { createContext } from "react";

// the createContext method returns a component, thats why
// we declare it with a capital letter at first, we will
// use this context which will cover all the routes of our
// webapp to transfer certain information across the components,
// in this case to handle private and public routes within the
// page
export const AuthContext = createContext();