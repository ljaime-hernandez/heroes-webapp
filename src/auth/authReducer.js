import { types } from "../types/types";

// the reducer should be a pure function in a manner on which everything
// should be resolved within it, the reducer should not exit to check on
// additional functions but only to work in itself and give an outcome no
// matter what arguments it receive, in this case our default return is 
// an empty object which is a default value for the state
export const authReducer = (state = {}, action) => {

    switch (action.type) {
        case types.login:
                return {
                    ...action.payload,
                    logged: true
                }

        case types.logout:
                return {
                    logged: false
                }

        default:
            return state;
    }

}