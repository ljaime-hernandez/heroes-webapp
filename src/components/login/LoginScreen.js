import { useContext } from 'react';
import {useNavigate} from 'react-router-dom';
import { AuthContext } from '../../auth/authContext';
import { types } from '../../types/types';

export const LoginScreen = () => {

    // we use the useNavigate method from the react-router-dom library to assign its content on a 
    // constant which we will later use for a click event
    const navigate = useNavigate();
    // we destructure the dispatch function of our context to send its arguments to our reducer
    const {dispatch} = useContext(AuthContext);

    const handleLogin = () => {

        // the login property of the types file is added to this object, along with a default
        // value which will be rendered on the navbar, the dispatch will then confirm the log
        // in, adding the values to the reducer object which will then be used by the context
        // for route handling in both the public and private components
        const action = {
            type: types.login,
            payload: {name: 'Miguel'}
        }

        dispatch(action);

        // the private route will always be updating the information of the last page visited
        // by retrieving the users location with a useLocation hook, we then retrieve it with
        // this getItem method, which, in case the user deletes or clears the cache of its browser,
        // will then assign the marvel path to this constant value by default
        const lastPath = localStorage.getItem('lastPath') || '/marvel';

        // once the button is clicked, we will send the user to the marvel page (by default, if theres a path
        // saved in the localStorage 'lastPath' item then we will use it), rendering its contents.
        // the 'replace: true' is used so if the user tries to return to the previous page, he is unable
        // to do so, we do this so once the user is logged out, he cannot come back until he login again,
        // not very specific on this example but conditionally makes sense on a session context
        navigate(lastPath, {
            replace: true,
        });
    }

    return (
        <div className='container mt-5'>
            <h1>Login Screen</h1>
            <hr/>

            <button 
                className='btn btn-primary'
                onClick={handleLogin}
            >
                Login
            </button>
        </div>
    )
}
