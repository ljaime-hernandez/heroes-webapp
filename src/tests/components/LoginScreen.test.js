import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import { AuthContext } from '../../auth/authContext';
import { LoginScreen } from '../../components/login/LoginScreen';
import { types } from '../../types/types';

/* to run this test:
1. run the 'npm install' command from the heroes-webapp folder 
2. run the 'npm install --save-dev enzyme' command (if you have not done so)
3. run the 'npm install --save-dev enzyme-to-json' command (if you have not done so)
4. run the 'npm install --save-dev @wojtekmaj/enzyme-adapter-react-17 --legacy-peer-deps' command (if you are using React 17 as i do)
5. run the 'npm install --save-dev @testing-library/react-hooks' command (if you have not done so)
6. make sure the setupTests.js file include the enzyme, enzyme-to-json and the react adapter libraries
7. run the command 'npm run test'
8. to have a clearer view of this single js test file, press p. then type the file name 'LoginScreen.test.js'
*/

const dispatch = jest.fn();
const mockNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate
}));

describe('tests on LoginScreen component', () => {

    test('should render properly by default', () => {
     
        // default value on the LoginScreen component, the context should have an user
        // in logout state and should pass a dispatch function from our custom reducer,
        // in this case the dispatch will be just a mock we will use just to confirm
        // if the function was called or not and with which parameters was it called
        const contextValue = {
            dispatch,
                user: {
                    logged: false
                }
            }

        const wrapper = mount(
            <MemoryRouter initialEntries={['/login']}>
                <AuthContext.Provider value={contextValue}>
                    <LoginScreen/>
                </AuthContext.Provider>
            </MemoryRouter>
            )

        expect(wrapper).toMatchSnapshot();
    });
    
    test('should call dispatch and navigate', () => {
        
        // we start again with a default logout user
        const contextValue = {
            dispatch,
                user: {
                    logged: false
                }
            }
        
        const wrapper = mount(
            <MemoryRouter initialEntries={['/login']}>
                <AuthContext.Provider value={contextValue}>
                    <LoginScreen/>
                </AuthContext.Provider>
            </MemoryRouter>
            )

        // we simulate the click function inside the LoginScreen component
        const handleClick = wrapper.find('button').prop('onClick');
        handleClick();

        // the action object we have by default inside our handleClick function
        // should change the state from logout to login in the user value from the
        // context, the dispatch would be the reducer function we use for it to change
        // its state, as the dispatch is a mock in this situation, we just have to confirm
        // the actual action object was sent as parameter accordingly
        expect(contextValue.dispatch).toHaveBeenCalledWith({
            type: types.login,
            payload: {
                name: 'Miguel'
            }
        });

        // the mockNavigate should be also called inside the handleClick with our default
        // parameter, which is supposed to be the marvel path by default, next we will
        // test the same function being called with the information stored in the
        // localStorage 'lastPath' item
        expect(mockNavigate).toHaveBeenCalledWith("/marvel", {"replace": true});

        localStorage.setItem('lastPath', '/dc');
        handleClick();
        expect(mockNavigate).toHaveBeenCalledWith("/dc", {"replace": true});

    });
    
})