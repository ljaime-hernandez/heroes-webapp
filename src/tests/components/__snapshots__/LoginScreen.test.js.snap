// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests on LoginScreen component should render properly by default 1`] = `
<div
  className="container mt-5"
>
  <h1>
    Login Screen
  </h1>
  <hr />
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Login
  </button>
</div>
`;
